### PR TITLE
TransformerEngine integration

### PR DESCRIPTION
**Pull request by carmocca**
- Implements padding support directly in the model.
- Adds a version of the model fully integrated with TransformerEngine
- Adds a guide for training with TransformerEngine

### CODE REVIEW

File: lit_parrot/model.py [hunk 1]
```diff
@@ -227,14 +264,20 @@ def forward(
...
+            padding = T - input_pos.size(0)
+            input_pos = torch.cat(
+                (input_pos, torch.full((padding,), padding_idx, device=input_pos.device, dtype=input_pos.dtype))
+            )
             k = cache_k.index_copy(2, input_pos, k)
             v = cache_v.index_copy(2, input_pos, v)
             kv_cache = k, v
 
-        # efficient attention using Flash Attention CUDA kernels
-        y = F.scaled_dot_product_attention(
```
Comment (carmocca) [inline lit_parrot/model.py:235]: I was getting nans using `scaled_dot_product_attention` with padding. I will try to repro it and report it to PyTorch

File: lit_parrot/model.py [hunk 2]
```diff
@@ -221,18 +241,28 @@ def forward(
...
-            q, k, v, attn_mask=mask, dropout_p=0.0, scale=1.0 / math.sqrt(self.config.head_size)
-        )
+        scale = 1.0 / math.sqrt(self.config.head_size)
+        if padding == 0:
+            y = F.scaled_dot_product_attention(q, k, v, mask, dropout_p=0.0, scale=scale)
+        else:
+            # `scaled_dot_product_attention` produces nans when padding is added
+            # TODO: create repro and report to pytorch
```
Comment (carmocca) [inline lit_parrot/model.py:261]: Done: https://github.com/pytorch/pytorch/issues/103749

```suggestion
            # `scaled_dot_product_attention` produces nans when padding is added
            # https://github.com/pytorch/pytorch/issues/103749
```

File: howto/fp8.md [hunk 3]
```diff
@@ -0,0 +1,119 @@
+# FP8 mixed precision
+
+[NVIDIA's Transformer Engine](https://github.com/NVIDIA/TransformerEngine) (TE) is a library for accelerating Transformer models on NVIDIA GPUs, including using 8-bit floating point (FP8) precision on Hopper GPUs, to provide better performance with lower memory utilization in both training and inference
+
+To use it, you'll need access to an H100 machine. If you don't have it, you can use CoreWeave's cloud.
```
Comment (Andrei-Aksionov) [inline howto/fp8.md:5]: I believe [Lambda](https://lambdalabs.com/) also provides H100.

File: lit_gpt/model_transformer_engine.py [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+        self,
+        idx: torch.Tensor,
+        max_seq_length: Optional[int] = None,
+        input_pos: Optional[torch.Tensor] = None,
+        padding_multiple: int = 8,
+    ) -> torch.Tensor:
+        B, T = idx.size()
+        use_kv_cache = input_pos is not None
+
+        T_padded = find_multiple(T, padding_multiple)
```
Comment (carmocca) [inline lit_gpt/model_transformer_engine.py:66]: This assumes that all sequences in the batch are good up to T, which might not be true for batched inference

**Comment by MF-FOOM** [comment]
oof, what's the reason for closing this?

**Comment by carmocca** [comment]
The current state of the PR is stale. I need to redo the version with all layers replaced.

Also, I won't add support for automatic padding again since it's only relevant for inference, and inference with TE wasn't worth it from my experiments.
